### PR TITLE
Revert using blockchainhook

### PR DESCRIPTION
**Pull request by matei-p**
**parent-child transfer and execute with success**
	parent	= -esdtValue
	child		= +esdtValue
	
**parent-child transfer and execute with failure**
	parent	= esdtValue
	child		= 0
	
**parent-child transfer failure**
	parent	= esdtValue 
	child		= 0

**parent-child transfer and execute with failure after burn builtin call (in child)** 
	parent	= esdtValue 
	child		= 0

**parent-child transfer and async execute with success** 
	parent	= -esdtValue
	child		= +esdtValue

**parent-child transfer and async execute with failure**
	parent	= esdtValue
	child		= 0

**parent-child transfer and async execute with success + callback failure** 
	parent	= -esdtValue
	child		= +esdtValue
	
**parent-child transfer and async execute with success + callback child-parent transfer**
 	parent	= -esdtValue +callbackTransfer
	child		= +esdtValue -callbackTransfer
 	
**parent-child transfer and async execute with success + callback child-parent transfer failed** 
	parent	= -esdtValue
	child		= +esdtValue	

**parent-child transfer and async execute with success + callback child-parent transfer success but callback fails** 
	parent 	= -esdtValue
	child		= +esdtValue

### CODE REVIEW

File: arwen/host/execution.go [hunk 1]
```diff
@@ -202,25 +202,28 @@ func copyTxHashesFromContext(copyEnabled bool, runtime arwen.RuntimeContext, inp
 func (host *vmHost) ExecuteOnDestContext(input *vmcommon.ContractCallInput) (vmOutput *vmcommon.VMOutput, asyncInfo *arwen.AsyncContextInfo, err error) {
 	log.Trace("ExecuteOnDestContext", "caller", input.CallerAddr, "dest", input.RecipientAddr, "function", input.Function)
 
-	builtinFunctionExecuted := false
 	scExecutionInput := input
 
+	blockchain := host.Blockchain()
```
Comment (sasurobert) [inline arwen/host/execution.go:207]: we have to delete any user of RevertESDT - there are 2 public functions. those must deleted

File: go.sum [hunk 2]
```diff
@@ -65,8 +66,9 @@ github.com/ElrondNetwork/elrond-go v1.1.38-0.20210322140841-6a27c533e8a5/go.mod
 github.com/ElrondNetwork/elrond-go v1.1.42/go.mod h1:mxhdRRV7PY+IHzznJ3MnX5/2YSMNxl2znD6f6Y8nuCM=
 github.com/ElrondNetwork/elrond-go v1.1.51/go.mod h1:0gT8uLRXSqmKDrrn55aXTSca0XGpg8aszx+Q5OPMNpY=
 github.com/ElrondNetwork/elrond-go v1.1.59-0.20210526130950-2a93e2e11c39/go.mod h1:0aZfahOLX4vO+1D6Ejx8XwOLTzd7pfcx94xzAOpt2bs=
-github.com/ElrondNetwork/elrond-go v1.1.60-0.20210528095714-a120ac08c5a7 h1:pyC2kblnQcqpEssAPFUp0uvUbHGloDDilf0nxRmMDao=
 github.com/ElrondNetwork/elrond-go v1.1.60-0.20210528095714-a120ac08c5a7/go.mod h1:t6uL0FdjyWPulWVnYaj6TgV3AHLw3LCiZvI9FQoORpU=
+github.com/ElrondNetwork/elrond-go v1.2.1-0.20210531130414-6ceb2ade721d h1:3/diuAqRtbVSE8twOjfnDQrhOrnpiiIyDSnVzcDfJvo=
```
Comment (sasurobert) [inline go.sum:70]: go mod tidy

File: ipc/arwenpart/blockchainGateway.go [hunk 3]
```diff
@@ -1,9 +1,10 @@
 package arwenpart
 
 import (
+	"github.com/ElrondNetwork/elrond-go/data/esdt"
```
Comment (sasurobert) [inline ipc/arwenpart/blockchainGateway.go:4]: sort imports

File: ipc/arwenpart/blockchainGateway.go [hunk 4]
```diff
@@ -442,11 +437,85 @@ func (blockchain *BlockchainHookGateway) GetCompiledCode(codeHash []byte) (bool,
...
+
+	response := rawResponse.(*common.MessageBlockchainGetESDTTokenResponse)
+	return response.Result, response.GetError()
 }
 
-// IsInterfaceNil returns true if underlying implementation is nil
+// IsInterfaceNil forwards a message to the actual hook
 func (blockchain *BlockchainHookGateway) IsInterfaceNil() bool {
-	return blockchain == nil
+
```
Comment (sasurobert) [inline ipc/arwenpart/blockchainGateway.go:475]: this is wrong. There is no need for IPC messaging for IsInterfaceNil. the only code here should be return blockchain == nil

File: mock/context/blockChainHookStub.go [hunk 5]
```diff
@@ -245,3 +247,19 @@ func (b *BlockchainHookStub) ClearCompiledCodes() {
 func (b *BlockchainHookStub) IsInterfaceNil() bool {
 	return b == nil
 }
+
+// GetSnapshot mocked method
```
Comment (sasurobert) [inline mock/context/blockChainHookStub.go:251]: should be moved before IsInterfaceNIl

File: arwen/host/execution.go (hunk 1 above)
Comment (sasurobert) [reply arwen/host/execution.go:207]: you have to remove host.RevertESDTTransfer(input) from TransferAndExecute function. Now you revert it through snapshot. after that you send tokens again.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: arwen/host/execution.go (hunk 1 above)
Comment (sasurobert) [reply arwen/host/execution.go:207]: the same happens on asnycCall with revert -> you make a revert snasphot here, but you make and actual transfer later - which is clearly wrong.

File: go.sum (hunk 2 above)
Comment (matei-p) [reply go.sum:70]: did not produce any change in go.sum

File: arwen/elrondapi/elrondei.go [hunk 6]
```diff
@@ -1043,11 +1043,7 @@ func TransferESDTNFTExecuteWithHost(
 		_, _, executeErr = host.ExecuteOnDestContext(contractCallInput)
 		if executeErr != nil {
 			logEEI.Trace("ESDT post-transfer execution failed", "error", executeErr)
-			_, _, revertErr := host.ExecuteESDTTransfer(sender, dest, tokenIdentifier, uint64(nonce), big.NewInt(0).SetBytes(valueBytes), vmcommon.AsynchronousCallBack, true)
-			if revertErr != nil {
-				logEEI.Warn("ESDT revert failed - forced fail execution for context", "error", executeErr)
-				_ = arwen.WithFaultAndHost(host, arwen.WrapError(executeErr).WrapWithError(revertErr), true)
-			}
+			_ = arwen.WithFaultAndHost(host, arwen.WrapError(executeErr).WrapWithError(executeErr), true)
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1046]: no arwenWithFault here.

File: arwen/host/asyncCall.go [hunk 7]
```diff
@@ -103,15 +108,15 @@ func (host *vmHost) determineAsyncCallExecutionMode(asyncCallInfo *arwen.AsyncCa
...
+	functionName, args, err := argParser.ParseData(string(asyncCallInfo.Data))
 	if err != nil {
 		return arwen.AsyncUnknown, err
 	}
 
 	sameShard := host.AreInSameShard(runtime.GetSCAddress(), asyncCallInfo.Destination)
 	if host.IsBuiltinFunctionName(functionName) {
 		if sameShard {
-			isESDTTransfer := functionName == core.BuiltInFunctionESDTTransfer || functionName == core.BuiltInFunctionESDTNFTTransfer
+			isESDTTransfer, _, _ := isESDTTransferOnReturnDataFromFunctionAndArgs(functionName, args)
```
Comment (sasurobert) [inline arwen/host/asyncCall.go:119]: good refactor.

**Review by sasurobert — APPROVED** [review]

File: arwen/contexts/blockchain.go [hunk 8]
```diff
@@ -328,3 +328,11 @@ func (context *blockchainContext) PopDiscard() {
 
 	context.stateStack = context.stateStack[:stateStackLen-1]
 }
+
+func (context *blockchainContext) GetSnapshot() int {
```
Comment (sasurobert) [inline arwen/contexts/blockchain.go:332]: add comment to public function

File: arwen/contexts/blockchain.go [hunk 9]
```diff
@@ -328,3 +328,11 @@ func (context *blockchainContext) PopDiscard() {
...
+	return context.blockChainHook.GetSnapshot()
+}
+
+func (context *blockchainContext) RevertToSnapshot(snapshot int) {
```
Comment (sasurobert) [inline arwen/contexts/blockchain.go:336]: add comment to public function

**Review by sasurobert — APPROVED** [review]

**Review by camilbancioiu — APPROVED** [review]

**Pull request merged**